LIPS/LIP-draft_on_coverage_app.md

```diff
@@ -0,0 +1,465 @@
+- request to cover slashing event (e.g. decreasing of the total pooled ETH amount between the two consecutive oracle reports)
+- request to burn shares for any other cases (non-cover)
+
+The proposed contract has two separate counters for the burnt shares: cover and non-cover ones. The contract should have exclusive access to the stETH shares burning. Also, it's highly desirable to only allow burning stETH from the contract's own balance.
+
+Finally, `LidoStETHBurner` provides logs and public getters to provide external access for proper accounting and monitoring. 
+
+### Sending a burn request
+
+The user sends a burn request transaction to the `LidoStETHBurner` contract by providing some stETH tokens and indicating the burning request type: cover or non-cover. The contract locks the provided stETH amount on its own balance and marks this amount for future burning.
+
+The contract also contains a pair of internal counters to collect the total amount to be burnt upon the next oracle report.
+
+### Burning execution on oracle report
+
+The contract enacts pending burning requests as part of the next oracle report by burning all associated stETH tokens from its own balance. After that, the contract increments burnt shares counters (cover and non-cover burnt shares are counted separately) and emits events.
+
+The contract tracks `LidoOracle` reports by implementing the `IBeaconReportReceiver` interface for a callback invocation (thanks to the implemented [LIP-2 Oracle upgrade](https://github.com/lidofinance/lido-improvement-proposals/blob/develop/LIPS/lip-2.md)).
+
+The main reason for performing token burn exactly as part of a Lido oracle update is to preserve a predictable and well-known rebase period since **_every_** burning event also leads to token rebase. External integrations and oracles may rely on the existing rebase period (which is about one-day).
+
+#### Gas consumption considerations
+
+Each quorum-reaching beacon report currently costs around 461000 gas.
+
+Adding the mechanism described in this proposal would increase the consumption as follows (with a 20% accuracy margin): by 720 gas for the case when there are no pending burn requests (base recurrent case); by 110000 gas when any number of burn requests of a single type is to be enacted (either cover or non-cover); by 160000 gas in the case when both cover a non-cover requests are to be enacted.
+
+This represents **less than a 0.2% increase for the base case** when no burn requests are pending, a 24% increase when applying one type of burn requests, and 35% increase when applying both types of burn requests.
+
+### Shares burnt counter
+
+We propose to store the total amount of shares ever burnt, distinguishing cover and non-cover shares, by maintaining two separate counters inside the `LidoStETHBurner` contract: `totalCoverSharesBurnt` and `totalNonCoverSharesBurnt`. The counters are increased when actual stETH burn is performed as part of the Lido Oracle report.
```

Comment:
I like the idea to have a counter for `totalNonCoverSharesBurnt` but what's the exact case of using this counter?

Reply:
I don't know the exact cases of using this counter for a particular reason too. :grin: 

But, in general, I think this way we preserve protocol transparency. Anyone can see the ever burnt shares amount (e.g. ```totalCoverSharesBurnt + totalNonCoverSharesBurnt```).

Also it could be used for stats and dashboards, for example, to explore `stETH`'s BME.

Reply:
Yeah, having the total number of shares ever burnt seems like a good idea in general since it might help in building future queries and dashboards. I'd ask this the other way: given that implementing the total counter is trivial, what are the reasons for not doing this?